query-planner-js/src/composedSchema/buildComposedSchema.ts

```diff
@@ -148,8 +150,11 @@ export function buildComposedSchema(document: DocumentNode): GraphQLSchema {
       type.astNode,
     );
 
+    // The assertion here guarantees the safety of the type cast below
+    // (typeMetadata as FederationEntityTypeMetadata). Adjustments to this assertion
+    // should account for this dependency.
     assert(
-      !(typeMetadata.isValueType && typeDirectivesArgs.length >= 1),
+      !(isValueTypeMetadata(typeMetadata) && typeDirectivesArgs.length >= 1),
```

Comment:
Almost every call to `isValueTypeMetadata` immediately `!`s it, and I think this one could be reasonably written as

```
!isValueTypeMetadata(typeMetaData) || typeDirectivesArgs.length === 0
```

so I'd suggest making the above change and then replacing isValueTypeMetadata with isEntityTypeMetadata?

That means less `!` overall, and specifically makes understanding the `as` below easier for me, since the assertion becomes "either it's the type we're going to `as` or the list is empty" which seems more straightforward than "it's not the case that it's both the type we're not mentioning in the `as` and the list is non-empty".

Reply:
Cool, glad you liked the direction I went. And I definitely agree with your suggestion! 77842761